### PR TITLE
[OSD-21095] Implements a new feature to allow supported opt-in region enablement during account creation process

### PR DESCRIPTION
**Pull request by reedcort**
# What is being added?
Implements a new feature to allow supported opt-in region enablement during account creation process

## Checklist before requesting review

- [x] I have tested this locally
- [x] I have included unit tests
- [ ] I have updated any corresponding documentation

## Steps To Manually Test

1. Start the operator
2. Run "make predeploy"
3. Set "feature.opt_in_regions" to true in hack/templates/aws.managed.openshift.io_v1alpha1_configmap.tmpl
4. Add  "opt-in-regions" entry to hack/templates/aws.managed.openshift.io_v1alpha1_configmap.tmpl
5. Run "make create-account"
6. Set account.Spec.accountPool to "hs-zero-size-accountpool"
7. Verify account CR state changes to READY
8. Verify in the AWS console that the opt-in regions specified in the configMap are enabled

Ref [OSD-21095](https://issues.redhat.com//browse/OSD-21095)

### CODE REVIEW

File: api/v1alpha1/account_types.go
Comment (iamkirkbater) [inline api/v1alpha1/account_types.go:112]: I'd advise against having this in the code, especially in the API for a few reasons:

1. Every change to add a new AWS region will now require a code change and promotion
2. Being in the API package, which other services (hive, ocm, etc) consume - we could run into issues with downstream consumers not updating, and if we ever make changes for any reason we'll have to run through deprecation procedures, etc.

Mostly - it's a large maintenance cost to add these to the code, and we'll _also_ need subsequent SOP updates to track these changes for our future selves.

File: controllers/account/account_controller.go
Comment (iamkirkbater) [inline controllers/account/account_controller.go:130]: Probably want to add log statements before these two return statements

File: controllers/account/account_controller.go
Comment (iamkirkbater) [inline controllers/account/account_controller.go:149]: Can we rename this var to something more descriptive?

Maybe this whole function could be abstracted?

File: controllers/account/account_controller.go
Comment (iamkirkbater) [inline controllers/account/account_controller.go:142]: I recommend abstracting this block of code into its own function.

Additionally, I think maybe this code may be better executed only when we need to opt into the region, otherwise we'll be calling this function every time an AccountCR reconciles - which on staging/int hives will be thousands of iterations. So maybe move the invocation of this functionality towards the block at line 413?

File: controllers/account/account_controller.go
Comment (iamkirkbater) [inline controllers/account/account_controller.go:418]: I think we need to rethink how we get to this point - I think at the end of the AccountCreating step we should decide what the next state is and put the CR into that state directly, and then return the reconciler - then when it gets back to this point in the code it skips the Account Creation stuff and gets here and goes "oh, I'm AccountOptingInRegions - I'll do that handler" and returns from that, then when we're done handling the opt-in region enablement we tell that to put it to the next state, instead of relying on it to just fall through as we're doing here for the InitializeRegions.

If that doesn't make sense, let me know I'll be happy to explain.

**Review by iamkirkbater — CHANGES_REQUESTED** [review]

File: controllers/account/region_enablement.go
Comment (AlexVulaj) [inline controllers/account/region_enablement.go:152]: Can we replace this with an `errors.As` ? I know we use this structure elsewhere in the codebase but it can lead to mistakes with wrapped errors. `errors.As` traverses all of the wrapped errors while checking for a specific error type.

There are also a few other places that do this in this file as well. I'll leave it to your discretion.

**Comment by AlexVulaj** [comment]
/approve

One comment that's more about semantics than function - it's a change I'd like to see but wouldn't hold up the PR for it.

**Pull request merged**
